File: sdk/python/school_sis_sdk/services/students.py
"""
Students service for School SIS SDK
"""

from typing import Dict, List, Optional, Any
from ..client import SISClient
from ..models import Student, ListParams, PaginationInfo


class StudentsService:
    """Service for managing students"""
    
    def __init__(self, client: SISClient):
        self.client = client
    
    def list(self, params: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        """Get list of students with optional filtering and pagination"""
        if params is None:
            params = {}
        
        response = self.client.get('/students', params=params)
        return response
    
    def get(self, student_id: str) -> Student:
        """Get a specific student by ID"""
        data = self.client.get(f'/students/{student_id}')
        return Student(**data)
    
    def create(self, student_data: Dict[str, Any]) -> Student:
        """Create a new student"""
        data = self.client.post('/students', json=student_data)
        return Student(**data)
    
    def update(self, student_id: str, student_data: Dict[str, Any]) -> Student:
        """Update a student"""
        data = self.client.put(f'/students/{student_id}', json=student_data)
        return Student(**data)
    
    def patch(self, student_id: str, student_data: Dict[str, Any]) -> Student:
        """Partially update a student"""
        data = self.client.patch(f'/students/{student_id}', json=student_data)
        return Student(**data)
    
    def delete(self, student_id: str) -> None:
        """Delete a student"""
        self.client.delete(f'/students/{student_id}')
    
    def search(self, query: str, params: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        """Search students by query"""
        if params is None:
            params = {}
        params['search'] = query
        return self.list(params)
    
    def get_by_grade_level(self, grade_level: str, params: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        """Get students by grade level"""
        if params is None:
            params = {}
        params['gradeLevel'] = grade_level
        return self.list(params)
    
    def get_by_status(self, status: str, params: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        """Get students by status"""
        if params is None:
            params = {}
        params['status'] = status
        return self.list(params)
    
    def get_by_academic_program(self, academic_program: str, params: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        """Get students by academic program"""
        if params is None:
            params = {}
        params['academicProgram'] = academic_program
        return self.list(params)
    
    def bulk_create(self, students: List[Dict[str, Any]]) -> Dict[str, Any]:
        """Bulk create students"""
        return self.client.post('/students/bulk', json={'students': students})
    
    def bulk_update(self, updates: List[Dict[str, Any]]) -> Dict[str, Any]:
        """Bulk update students"""
        return self.client.put('/students/bulk', json={'updates': updates})
    
    def bulk_delete(self, student_ids: List[str]) -> Dict[str, Any]:
        """Bulk delete students"""
        return self.client.delete('/students/bulk', json={'ids': student_ids})
    
    def export_csv(self, params: Optional[Dict[str, Any]] = None) -> None:
        """Export students to CSV"""
        from datetime import datetime
        filename = f"students_{datetime.now().strftime('%Y-%m-%d')}.csv"
        self.client.download_file('/students/export/csv', filename)
        print(f"Students exported to {filename}")
    
    def import_csv(self, file_path: str) -> Dict[str, Any]:
        """Import students from CSV"""
        return self.client.upload_file('/students/import/csv', file_path)
    
    def get_statistics(self) -> Dict[str, Any]:
        """Get student statistics"""
        return self.client.get('/students/statistics')
    
    def get_academic_history(self, student_id: str) -> Dict[str, Any]:
        """Get student's academic history"""
        return self.client.get(f'/students/{student_id}/academic-history')
    
    def get_emergency_contacts(self, student_id: str) -> Dict[str, Any]:
        """Get student's emergency contacts"""
        return self.client.get(f'/students/{student_id}/emergency-contacts')
    
    def update_emergency_contacts(self, student_id: str, contacts: Dict[str, Any]) -> Student:
        """Update student's emergency contacts"""
        data = self.client.put(f'/students/{student_id}/emergency-contacts', json=contacts)
        return Student(**data)
    
    def get_medical_info(self, student_id: str) -> Dict[str, Any]:
        """Get student's medical information"""
        return self.client.get(f'/students/{student_id}/medical')
    
    def update_medical_info(self, student_id: str, medical_info: Dict[str, Any]) -> Student:
        """Update student's medical information"""
        data = self.client.put(f'/students/{student_id}/medical', json=medical_info)
        return Student(**data)
    
    def upload_document(self, student_id: str, file_path: str, document_type: str, description: Optional[str] = None) -> Dict[str, Any]:
        """Upload student document"""
        additional_data = {
            'documentType': document_type,
            'description': description
        }
        return self.client.upload_file(f'/students/{student_id}/documents', file_path, additional_data)
    
    def get_documents(self, student_id: str) -> List[Dict[str, Any]]:
        """Get student documents"""
        return self.client.get(f'/students/{student_id}/documents')
    
    def delete_document(self, student_id: str, document_id: str) -> None:
        """Delete student document"""
        self.client.delete(f'/students/{student_id}/documents/{document_id}')
    
    def upload_profile_picture(self, student_id: str, file_path: str) -> Dict[str, Any]:
        """Upload student profile picture"""
        return self.client.upload_file(f'/students/{student_id}/profile-picture', file_path)
    
    def get_family_members(self, student_id: str) -> List[Dict[str, Any]]:
        """Get student's family members (parents/guardians)"""
        return self.client.get(f'/students/{student_id}/family')
    
    def add_family_member(self, student_id: str, family_member: Dict[str, Any]) -> Dict[str, Any]:
        """Add family member to student"""
        return self.client.post(f'/students/{student_id}/family', json=family_member)
    
    def remove_family_member(self, student_id: str, family_member_id: str) -> None:
        """Remove family member from student"""
        self.client.delete(f'/students/{student_id}/family/{family_member_id}')

File: sdk/javascript/src/services/students.ts
import { SISClient } from '../client';
import { Student, ListParams, PaginationInfo } from '../types';

export class StudentsService {
  constructor(private client: SISClient) {}

  /**
   * Get list of students with optional filtering and pagination
   */
  async list(params: ListParams = {}): Promise<{ students: Student[]; pagination: PaginationInfo }> {
    const response = await this.client.get<{ students: Student[]; pagination: PaginationInfo }>('/students', params);
    return response;
  }

  /**
   * Get a specific student by ID
   */
  async get(id: string): Promise<Student> {
    return this.client.get<Student>(`/students/${id}`);
  }

  /**
   * Create a new student
   */
  async create(studentData: Partial<Student>): Promise<Student> {
    return this.client.post<Student>('/students', studentData);
  }

  /**
   * Update a student
   */
  async update(id: string, studentData: Partial<Student>): Promise<Student> {
    return this.client.put<Student>(`/students/${id}`, studentData);
  }

  /**
   * Partially update a student
   */
  async patch(id: string, studentData: Partial<Student>): Promise<Student> {
    return this.client.patch<Student>(`/students/${id}`, studentData);
  }

  /**
   * Delete a student
   */
  async delete(id: string): Promise<void> {
    return this.client.delete<void>(`/students/${id}`);
  }

  /**
   * Search students by query
   */
  async search(query: string, params: ListParams = {}): Promise<{ students: Student[]; pagination: PaginationInfo }> {
    return this.list({ ...params, search: query });
  }

  /**
   * Get students by grade level
   */
  async getByGradeLevel(gradeLevel: string, params: ListParams = {}): Promise<{ students: Student[]; pagination: PaginationInfo }> {
    return this.list({ ...params, gradeLevel });
  }

  /**
   * Get students by status
   */
  async getByStatus(status: string, params: ListParams = {}): Promise<{ students: Student[]; pagination: PaginationInfo }> {
    return this.list({ ...params, status });
  }

  /**
   * Get students by academic program
   */
  async getByAcademicProgram(academicProgram: string, params: ListParams = {}): Promise<{ students: Student[]; pagination: PaginationInfo }> {
    return this.list({ ...params, academicProgram });
  }

  /**
   * Bulk create students
   */
  async bulkCreate(students: Partial<Student>[]): Promise<{ students: Student[]; errors: any[] }> {
    return this.client.post<{ students: Student[]; errors: any[] }>('/students/bulk', { students });
  }

  /**
   * Bulk update students
   */
  async bulkUpdate(updates: Array<{ id: string; data: Partial<Student> }>): Promise<{ students: Student[]; errors: any[] }> {
    return this.client.put<{ students: Student[]; errors: any[] }>('/students/bulk', { updates });
  }

  /**
   * Bulk delete students
   */
  async bulkDelete(ids: string[]): Promise<{ deleted: string[]; errors: any[] }> {
    return this.client.delete<{ deleted: string[]; errors: any[] }>('/students/bulk', { data: { ids } });
  }

  /**
   * Export students to CSV
   */
  async exportCSV(params: ListParams = {}): Promise<Blob> {
    return this.client.downloadFile('/students/export/csv', `students_${new Date().toISOString().split('T')[0]}.csv`);
  }

  /**
   * Import students from CSV
   */
  async importCSV(file: File): Promise<{ imported: number; errors: any[] }> {
    return this.client.uploadFile<{ imported: number; errors: any[] }>('/students/import/csv', file);
  }

  /**
   * Get student statistics
   */
  async getStatistics(): Promise<{
    total: number;
    byGradeLevel: Record<string, number>;
    byStatus: Record<string, number>;
    byAcademicProgram: Record<string, number>;
    recentEnrollments: number;
    upcomingGraduations: number;
  }> {
    return this.client.get('/students/statistics');
  }

  /**
   * Get student's academic history
   */
  async getAcademicHistory(id: string): Promise<{
    student: Student;
    classes: any[];
    grades: any[];
    attendance: any[];
  }> {
    return this.client.get(`/students/${id}/academic-history`);
  }

  /**
   * Get student's emergency contacts
   */
  async getEmergencyContacts(id: string): Promise<{
    primary: any;
    secondary: any;
  }> {
    return this.client.get(`/students/${id}/emergency-contacts`);
  }

  /**
   * Update student's emergency contacts
   */
  async updateEmergencyContacts(id: string, contacts: {
    primary?: any;
    secondary?: any;
  }): Promise<Student> {
    return this.client.put<Student>(`/students/${id}/emergency-contacts`, contacts);
  }

  /**
   * Get student's medical information
   */
  async getMedicalInfo(id: string): Promise<{
    conditions: string;
    allergies: string;
    medications: string;
    insurance: string;
    physician: string;
  }> {
    return this.client.get(`/students/${id}/medical`);
  }

  /**
   * Update student's medical information
   */
  async updateMedicalInfo(id: string, medicalInfo: {
    conditions?: string;
    allergies?: string;
    medications?: string;
    insurance?: string;
    physician?: string;
  }): Promise<Student> {
    return this.client.put<Student>(`/students/${id}/medical`, medicalInfo);
  }

  /**
   * Upload student document
   */
  async uploadDocument(id: string, file: File, documentType: string, description?: string): Promise<{
    id: string;
    filename: string;
    url: string;
    documentType: string;
    description?: string;
  }> {
    return this.client.uploadFile(`/students/${id}/documents`, file, {
      documentType,
      description
    });
  }

  /**
   * Get student documents
   */
  async getDocuments(id: string): Promise<Array<{
    id: string;
    filename: string;
    url: string;
    documentType: string;
    description?: string;
    uploadedAt: string;
  }>> {
    return this.client.get(`/students/${id}/documents`);
  }

  /**
   * Delete student document
   */
  async deleteDocument(id: string, documentId: string): Promise<void> {
    return this.client.delete(`/students/${id}/documents/${documentId}`);
  }

  /**
   * Upload student profile picture
   */
  async uploadProfilePicture(id: string, file: File): Promise<{
    url: string;
    filename: string;
  }> {
    return this.client.uploadFile(`/students/${id}/profile-picture`, file);
  }

  /**
   * Get student's family members (parents/guardians)
   */
  async getFamilyMembers(id: string): Promise<Array<{
    id: string;
    relationship: string;
    firstName: string;
    lastName: string;
    email?: string;
    phone?: string;
    isPrimary: boolean;
  }>> {
    return this.client.get(`/students/${id}/family`);
  }

  /**
   * Add family member to student
   */
  async addFamilyMember(id: string, familyMember: {
    relationship: string;
    firstName: string;
    lastName: string;
    email?: string;
    phone?: string;
    isPrimary?: boolean;
  }): Promise<any> {
    return this.client.post(`/students/${id}/family`, familyMember);
  }

  /**
   * Remove family member from student
   */
  async removeFamilyMember(id: string, familyMemberId: string): Promise<void> {
    return this.client.delete(`/students/${id}/family/${familyMemberId}`);
  }
}
